submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,70 @@
+## Support multiple cable drivers within the same supercluster
+
+## Summary
+
+Submariner supports several cable driver implementations (libreswan, wireguard et al) however only one can be 
+configured within the same supercluster, that is all clusters that join a broker must use the same cable driver.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+supercluster. For example, a supercluster could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+### User-facing API
+
+There needs to be a way for a user to specify their intent for two clusters to use a specific cable driver
+implementation and/or configuration. At a minimum, the intent must include identifying information for the two
+clusters and the cable driver implementation.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This following CRD is proposed:
+
+```
+appVersion: v1
+kind: ClusterLinkConfig
+metadata:
+  name: east-west-link
+spec:
+  leftCluster: 
+    id: east
+  rightCluster: 
+    id: west
+  cableDriver: 
+    name: libreswan
+    options:
+      encrypt: true
+```
+
+The `leftCluster` and `rightCluster` fields identify the two clusters. The simplest identifier is the cluster's unique
```

Comment:
So @tpantelis , in a setup where we have two Clusters in OnPrem and one in public Cloud. Do we have to create multiple ClusterLinkConfig CRDs on **each cluster**? 

And is it mandatory to create the CRDs?

If the CRD is not configured, is the assumption that encryption would be used?

Reply:
I think we should have a set of defaults, same as currently (strongSwan, ultimately Libreswan, with encryption by default), and only override that if a link CR is present. The link CRs could be created on the broker and synced to the appropriate clusters.

Reply:
Yes, it makes sense to assume encryption by default and override the configuration for specific links using the CRD.

Reply:
> I think we should have a set of defaults, same as currently (strongSwan, ultimately Libreswan, with encryption by default), and only override that if a link CR is present. 

yes - I was thinking the default driver  + config as we have now that is passed via the env would remain the same (at least initially).  

> The link CRs could be created on the broker and synced to the appropriate clusters.

yes. This is a rough first draft - I didn't mention any of these details yet.

Reply:
We could resolve this by adding a priority field in the link configs....

The default could be created with priority 100, ... but if we add higher priority ones "1..", they would override  (or inverse numbering, whatever makes more sense)